Allow fake models to run with ONNXProgram.__call__

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122435
* __->__ #122230
* #122196

In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

Commit:
Allow fake models to run with ONNXProgram.__call__

In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -705,25 +706,44 @@ def __call__(
+                    " to populate the model with initializers before being execution."
+                )
+                # TODO: Revisit the need of `model_with_state_dict` being a real model and not just its state
+                onnx_model = os.path.join(tmpdir_path, "model.onnx")
```
Not sure if I understand correctly. Is the scenario that, the weights are already serialized to somewhere, so the onnx model itself is very small, and all weight tensor protos are pointing to external data through a relative path?

And a few more questions:
- Are the weights stored in the same temp folder, so that serializing model there resolves the core restriction of model and weights being in the same folder?
- Why do we still need the following `adapt_torch_inputs_to_onnx` step which seemingly appends weights to model inputs?

thiagocrepaldi replied:
In this scenario, the model has no initializers (because it came from fake mode), but the initializers WILL be serialized into external data on a temporary folder. The API assumes - for now - that `enable_fake_mode` captured a checkpoint file during export (at `ONNXProgram.fake_context._model_state_dicts`). Models and initializers are saved in the same temp folder - and deleted after `__call__` returns.



Two things:
A) When fake mode is not used, `adapt_torch_inputs_to_onnx` is needed to append initializer to model. 
B) This PR is only making easy the life of user to run fake models. It ignores the scenari in  which users do want to specify initializers as input

All APIs will probably will need revision to better accomodate A) and B)

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

Commit:
Update on "Allow fake models to run with ONNXProgram.__call__"


In order to a fake model to run using ONNXProgram.__call__
interface, we need to save the model into disk along with external data
before executing the model. This is what this PR implements

An alternative is to ONNXProgram.__call__ to detect that the model
was exported with fake mode and explicit raise an exception when
ONNXProgram.__call__ is executed. The exception message would instruct
the user to call ONNXProgram.save and manually execute the model using
the ONNX runtime of choice.

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge

thiagocrepaldi commented:
@malfet @kit1980 Just FYI, pytorchbot has been refusing to merge PRs pointing CI failure that didn't happen. Second PR with this same symptom already

@pytorchbot merge -f "Flaky pytorchbot. Lists failed tasks that havent actually failed"
